src/Umbraco.Infrastructure/DistributedLocking/InMemoryDistributedLockingMechanism.cs

```diff
@@ -75,78 +77,94 @@ public IDistributedLock WriteLock(int lockId, TimeSpan? obtainLockTimeout)
-                    return InMemoryDistributedLock.Write(this, lockId);
+                    return distributedLock;
                 }
             }
 
             Thread.Sleep(50);
-        }
-        while (timer.Elapsed < obtainLockTimeout);
+        } while (timer.Elapsed < obtainLockTimeout);
 
         throw new DistributedWriteLockTimeoutException(lockId);
     }
 
-    private void DropLock(int lockId, DistributedLockType lockType)
+    private void DropLock(InMemoryDistributedLock distributedLock)
     {
         lock (_locks)
         {
-            _locks.DropLock(lockId, lockType);
-            _logger.LogDebug("Dropped {lockType} for id {id}", lockType, lockId);
+            _locks.DropLock(distributedLock);
+            _logger.LogDebug("Dropped {lockType} for id {id}", distributedLock.LockType, distributedLock.LockId);
         }
     }
 
     private class Locks
     {
-        private readonly IDictionary<int, int> _readLocks = new Dictionary<int, int>();
-        private readonly HashSet<int> _writeLocks = new();
+        private readonly InMemoryDistributedLockingMechanism _parent;
+        private readonly List<InMemoryDistributedLock> _internalLocks = new();
+
+        public Locks(InMemoryDistributedLockingMechanism parent)
+        {
+            _parent = parent;
+        }
+
 
-        public bool TrAcquireReadLock(int lockId)
+        public bool TrAcquireReadLock(int lockId, out InMemoryDistributedLock distributedLock)
         {
-            if (_writeLocks.Contains(lockId))
+            if (HasExistingWriteLock(lockId))
             {
+                distributedLock = null;
                 return false;
             }
 
-            _readLocks[lockId] = GetReaderCount(lockId) + 1;
+            distributedLock = InMemoryDistributedLock.Read(_parent, lockId);
+            _internalLocks.Add(distributedLock);
             return true;
         }
 
-        public bool TryAcquireWriteLock(int lockId)
+        public bool TryAcquireWriteLock(int lockId, out InMemoryDistributedLock distributedLock)
         {
-            if (GetReaderCount(lockId) > 0)
+            if (HasExistingWriteLock(lockId))
             {
+                distributedLock = null;
                 return false;
             }
 
-            if (_writeLocks.Contains(lockId))
+            // No one has any sort of lock for this ID, so go ahead.
+            if (_internalLocks.All(x => x.LockId != lockId))
             {
-                return false;
+                distributedLock = InMemoryDistributedLock.Write(_parent, lockId);
+                _internalLocks.Add(distributedLock);
+                return true;
             }
 
-            _writeLocks.Add(lockId);
-            return true;
+            var existingLocks = _internalLocks.Where(x => x.LockId == lockId).ToList();
+            if (existingLocks.Count == 1 && existingLocks.First().ThreadId == Thread.CurrentThread.ManagedThreadId)
+            {
```

Comment:
We can do better if we make scope reader -> writer upgrades explicit rather than implied (but explicit upgrade is noop for sql server implementation).

Also tried with ReaderWriiterLockSlim but it's a bit meh as we don't know whether we need an upgradeable lock in advance so we kill concurrency (only single upgradeable lock at a time).

tl;dr. I'm not particularly happy with this setup but it's "OK" for now.

Might want to revisit this being the default impl when none is set in config.